### ISSUE TITLE
I configured ops_to_register.h and set #define SHOULD_REGISTER_OP_GRADIENT false, but unexpectedly the generated libtensorflow_inference.so file get larger.

### ISSUE THREAD

**Issue by scucheri** [tensorflow/tensorflow#12096]
### Describe the problem
 I  configured ops_to_register.h as below:

#define SHOULD_REGISTER_OP(op) true
#define SHOULD_REGISTER_OP_GRADIENT false
#define SHOULD_REGISTER_OP_KERNEL(clz) true

set  SHOULD_REGISTER_OP_GRADIENT false. And then I used command below to generate libtensorflow_inference.so file. 

`bazel build -c opt --copt="-DSELECTIVE_REGISTRATION" \                                         
            --copt="-DSUPPORT_SELECTIVE_REGISTRATION" \
            //tensorflow/contrib/android:libtensorflow_inference.so \
            --host_crosstool_top=@bazel_tools//tools/cpp:toolchain \
            --crosstool_top=//external:android/crosstool --cpu=armeabi-v7a`

I expected that this configuration will reduce gradient operations and make the genereted libtensorflow_inference.so file smaller. But unexpectedly, the generated file is 16M ,while  the .so file generate by command below which will include all operators is 9.7M.

` bazel build -c opt //tensorflow/contrib/android:libtensorflow_inference.so \
         --crosstool_top=//external:android/crosstool \
         --host_crosstool_top=@bazel_tools//tools/cpp:toolchain \
         --cpu=armeabi-v7a`

### Question
I can't understand why the .so file get larger when  I set SHOULD_REGISTER_OP_GRADIENT false than the original .so file which include all operators. Could you please explain it to me ? Thank you very much and hope for your answer.

**Comment by andrewharp** [comment]
Selective registration draws from a larger set of potential ops than the regular TF build. It appears what you have done with your macro definitions is to set it to include all these ops, excluding only the gradients which are used for training. Gradients are already excluded by default, and you've added more ops, therefore it makes sense you'll get a larger resulting library.

As explained in #11936, you'll want to used [print_selective_registration_header.py](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/tools/print_selective_registration_header.py) to generate a custom header which only causes the ops you actually use to be linked in.

**Comment by scucheri** [comment]
Yeah , I can reduce the .so size when using method of print_selective_registration_header.py. But I want to make it available for different algorithm models.  

So the default configuration doesn't include all SHOULD_REGISTER_OP_KERNEL and SHOULD_REGISTER_OP opretions ? But I see the code of selective_registration.h id like this:
![image](https://user-images.githubusercontent.com/10495849/29102454-d011a392-7ceb-11e7-98c4-baa55c0e1d7d.png)

I get confused. Which file configured the default  included operations ? Is it selective_registration.h ?  Can I get all available operations somewhere ?

**Comment by andrewharp** [comment]
Those defines you are referencing are only active if you _have not_ defined [SELECTIVE_REGISTRATION](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/framework/selective_registration.h#L21). Otherwise they come from ops_to_register.h.

As described in the comments, print_selective_registration.py will give you an ops_to_register.h file.

However if you want a general-purpose library capable of handling multiple models, you should probably just stick to the default builds. The entire point of selective registration is to remove unnecessary ops, and if you want to support different models you need to know ahead of time which ones are safe to remove.

For further assistance on this I would recommend posting on StackOverflow, as there is a larger community for support there, thanks.

**Comment by scucheri** [comment]
Thank you very much~
